[✨feat] 스크랩 추가 API 구현

jsoonworld opened the pull request:
# 📄 Work Description  
- 스크랩 추가 API를 구현했습니다.  
- 인턴십 공고와 사용자 정보를 바탕으로 스크랩을 생성하며, 중복 스크랩 방지를 위해 사전 검증 로직을 포함했습니다.  
- 요청 DTO(`ScrapRequest`)에 color 값을 받으며, 대소문자 구분 없이 색상을 매핑할 수 있도록 개선했습니다.  
- 관련된 예외 케이스(중복 스크랩, 공고 없음, 사용자 없음)를 처리하는 `ScrapException`도 함께 적용했습니다.  
- 정적 팩토리 메서드(`of`, `from`)를 각 도메인 객체에 적용해 객체 생성을 명확히 했습니다.  

# 💭 Thoughts  
- `userId`를 임시로 하드코딩한 상태인데, 이후 인증 정보를 기반으로 `@AuthenticationPrincipal` 등으로 교체하는 작업이 필요합니다.  
- PR을 더 작은 단위로 나누고 싶었지만, API 하나를 처음부터 끝까지 구현하는 과정에서는 흐름을 이해하는 게 더 중요하다고 판단해 API 단위로 PR을 구성했어요. 유빈님은 어떻게 생각하시나요? 앞으로의 PR 전략에 대해서도 함께 이야기해보면 좋을 것 같아요! 😊  
- 구현하면서 흥미로웠던 점은 자바에서는 생성자 주입을 위해 `@RequiredArgsConstructor`를 사용했었는데, 코틀린에서는 명시적으로 애너테이션 없이도 바로 의존성 주입이 가능하다는 점이 새로웠어요.  
- 이제 각 서비스 클래스 안에 여러 API가 생기면, 메서드 수가 많아져 가시성에 영향을 줄 수 있다는 고민이 생겼어요.  
  그래서 하나의 API 메서드마다 인터페이스를 따로 만들어 응집도 있게 분리하는 것도 고려해봤는데요,  
  이렇게 되면 관심사 분리는 잘 되지만 반대로 인터페이스와 구현체가 너무 많아져서 협업 시 추상화 구조를 파악하기 어려울 수도 있겠다는 고민도 들었어요.  
  예를 들어, 스크랩 서비스에 API가 4개 있다면 각 기능별 인터페이스 4개와 구현체 4개가 생기는 구조인데요,  
  이런 구조가 협업과 유지보수에 도움이 될지, 혹은 오히려 복잡도를 높일지 같이 이야기 나눠보고 싶습니다! 🙌


# ✅ Testing Result  
![스크린샷](https://github.com/user-attachments/assets/9e3e4eb0-bd06-48f3-9c21-8fdfa15056e5)

# 🗂 Related Issue  
- closed #80

jsoonworld committed:
chore(build) : 테스트 관련 의존성 추가 및 AOT 설정 비활성화

- springmockk, mockk, spring-security-test 테스트 의존성 추가
- processTestAot 태스크 비활성화 처리로 AOT 테스트 빌드 오류 방지

jsoonworld committed:
feat(Auth) : 정적 팩토리 메서드 of() 추가 및 @JoinColumn 이름 수정

- user 필드 매핑을 위한 @JoinColumn name 속성을 'user_id'로 수정
- Auth 객체 생성을 위한 of 정적 팩토리 메서드 구현

jsoonworld committed:
feat(AuthRepository) : Auth 엔티티용 JPA 리포지토리 인터페이스 생성

- Auth 엔티티를 관리하기 위한 AuthRepository 인터페이스 정의

jsoonworld committed:
feat(InternshipAnnouncement) : 정적 팩토리 메서드 of() 추가

- InternshipAnnouncement 객체 생성을 위한 of 정적 팩토리 메서드 구현
- 선택적 필드를 포함하여 유연하게 생성 가능하도록 구성

jsoonworld committed:
feat(InternshipAnnouncementRepository) : 인턴십 공고 리포지토리 인터페이스 생성

- InternshipAnnouncement 엔티티 관리를 위한 JPA 리포지토리 정의

jsoonworld committed:
feat(ScrapRepository) : 스크랩 여부 확인 메서드 추가

- Scrap 엔티티용 JPA 리포지토리 인터페이스 생성
- 사용자와 인턴십 공고 ID를 기반으로 스크랩 존재 여부 확인 메서드 추가

jsoonworld committed:
refactor(Color) : 색상 매칭 시 대소문자 구분 없이 비교하도록 개선

- Color.from 메서드에서 입력 문자열을 대소문자 구분 없이 처리
- 사용자 입력 유연성 향상

jsoonworld committed:
feat(User) : 정적 팩토리 메서드 of() 추가 및 생성자 접근 제어

- User 객체 생성을 위한 of 정적 팩토리 메서드 도입
- 생성자에 private 접근 제어자를 추가해 생성 제어 강화

jsoonworld committed:
feat(UserRepository) : User 엔티티용 JPA 리포지토리 인터페이스 생성

- User 엔티티 관리를 위한 기본 CRUD 리포지토리 정의

jsoonworld committed:
feat(UserName) : 정적 팩토리 메서드 from() 추가

- UserName 객체 생성을 위한 정적 팩토리 메서드 from 구현
- 생성 로직을 명시적으로 표현하여 가독성 및 확장성 향상

jsoonworld committed:
feat(ScrapErrorCode) : 스크랩 및 유저 관련 예외 코드 추가

jsoonworld committed:
feat(ScrapRequest) : 스크랩 요청을 위한 DTO 클래스 추가

- 클라이언트로부터 전달받은 스크랩 색상 정보를 담기 위한 ScrapRequest DTO 생성

jsoonworld committed:
feat(ScrapService) : 스크랩 생성 서비스 로직 구현

- 유저와 인턴십 공고 존재 여부 검증
- 중복 스크랩 방지 예외 처리 추가
- 요청된 색상으로 Scrap 객체 생성 및 저장
- 공고의 스크랩 수 증가 처리 포함

jsoonworld committed:
feat(ScrapController) : 스크랩 생성 API 구현

- POST /api/v1/scraps/{internshipAnnouncementId} 엔드포인트 추가
- 요청 본문으로 ScrapRequest 받기
- 임시 userId를 사용해 ScrapService 호출 및 응답 처리

jsoonworld committed:
chore(config) : 테스트용 H2 DB 설정 변경 및 ddl-auto 옵션 수정

- H2 DB에 MySQL 모드 및 DB_CLOSE_DELAY=-1 옵션 추가
- ddl-auto 설정을 create-drop → none으로 변경하여 자동 테이블 생성 방지

jsoonworld committed:
test(UserTest) : User 객체 생성 방식 변경에 따른 테스트 코드 수정

- 기존 생성자 직접 호출에서 정적 팩토리 메서드 of() 방식으로 변경
- profile 및 userState 값 전달 방식 반영

jsoonworld committed:
test(ScrapServiceTest) : 스크랩 서비스 로직 단위 테스트 추가

- 중복 스크랩 예외 처리 테스트
- 공고/사용자 미존재 예외 처리 테스트
- 정상적인 스크랩 저장 및 스크랩 수 증가 검증

jsoonworld committed:
test(ScrapControllerTest) : 스크랩 생성 API 테스트 추가

- /api/v1/scraps/{internshipAnnouncementId} 엔드포인트에 대한 POST 요청 테스트
- ScrapService mock 처리 및 성공 시 201 Created 응답 검증

jsoonworld committed:
chore(dependencies) : H2 DB 드라이버 의존성을 runtime 환경에서 사용하도록 수정

jsoonworld committed:
refactor(UserName) : 문자열 템플릿 포함된 에러 메시지를 const val → val로 수정

- Kotlin에서는 const val에 문자열 템플릿 사용이 불가능하므로 ERROR_LENGTH를 val로 변경
- 컴파일 에러 예방 및 코드 안정성 확보

jsoonworld committed:
Merge branch 'develop' into feat/#80

jsoonworld committed:
test(ScrapServiceTest) : 매직 리터럴 제거를 위해 Color enum 사용

jsoonworld commented on src/test/kotlin/com/terning/server/kotlin/application/scrap/ScrapServiceTest.kt:
```diff
@@ -0,0 +1,99 @@
+
+        verify { announcement.increaseScrapCount() }
+
+        assertEquals("#4AA9F2", scrapSlot.captured.hexColor())
```
피드백 감사합니다! 😊
말씀 주신 부분을 반영해서 매직 리터럴이었던 "#4AA9F2"를 Color.BLUE.toHexString()으로 변경했습니다.

이렇게 수정하니 "BLUE"라는 입력값과 기대 결과(hex 코드) 간의 매핑이 테스트 코드 안에서도 더 명확하게 드러나는 것 같아요. 덕분에 도메인 지식을 반영한 코드로 개선할 수 있었습니다!
꼼꼼한 리뷰 감사드립니다 🙏

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/application/scrap/ScrapService.kt:
```diff
@@ -0,0 +1,45 @@
+                .orElseThrow { ScrapException(ScrapErrorCode.USER_NOT_FOUND) }
+
+        val color = Color.from(request.color)
+        val scrap = Scrap.of(user, announcement, color)
```
아래처럼 named argument를 사용하면 코드 가독성이 좋아질 것 같아요! 어떻게 생각하시나요?
```suggestion
        val scrap = Scrap.of(
            user = user, 
            internshipAnnouncement = announcement, 
            color = color
        )
```

jsoonworld replied:
WOWOWO... 이거군요! 😮

확실히 네이밍 아규먼트를 사용하니까 가독성이 훨씬 좋아지네요.
약간의 중복처럼 느껴질 수도 있지만, 실수를 줄일 수 있고, 특히 협업할 때 의도가 명확히 드러나서 더 좋은 것 같아요!
말씀 주신 방향대로 개선하겠습니다 :) 💪

다만 한 가지 고민되는 점은,
변수명이 이미 명확할 경우 네이밍 아규먼트를 추가하는 것이 오히려 중복처럼 느껴질 수 있다는 점이에요.
특히 생성자나 팩토리 메서드의 매개변수 순서가 명확할 때는 코드가 간결하지 않아 보일 수 있다는 아쉬움도 있습니다.

그럼에도 불구하고 저희가
- 2개 이상 인자가 전달되는 경우에는 네임드 아규먼트를 사용하자고 정했고,
-  전반적인 통일성을 가져가는 게 협업과 유지보수 측면에서 훨씬 유리하다고 판단되기 때문에

이번에도 네임드 아규먼트 방식으로 반영하겠습니다! 😊
꼼꼼 피드백 감사합니다! 🙌

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/user/vo/UserName.kt:
```diff
@@ -22,6 +22,10 @@ class UserName(
-        private val ERROR_LENGTH = "이름은 ${MIN_LENGTH}~${MAX_LENGTH}자여야 합니다."
+        private val ERROR_LENGTH = "이름은 $MIN_LENGTH~${MAX_LENGTH}자여야 합니다."
+
+        fun from(value: String): UserName {
+            return UserName(value)
+        }
```
여기두.. 표현식이 좋을 것 같아요..!
```suggestion
        fun from(value: String): UserName = UserName(value)
```

jsoonworld replied:
아이코! 이 부분은 놓치고 있었네요 😅
(아직 자바 인간인지... 표현식이 아닌 걸 봐도 어색하다는 생각이 잘 안 드네요 ㅠㅠ)
말씀 주신 대로 바로 반영하겠습니다!
꼼꼼하게 챙겨주셔서 정말 감사해요 🙏✨

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ScrapController.kt:
```diff
@@ -0,0 +1,38 @@
+        @PathVariable internshipAnnouncementId: Long,
+        @RequestBody scrapRequest: ScrapRequest,
+    ): ResponseEntity<ApiResponse<Unit>> {
+        val userId: Long = 1 // 임시 userId
```
주석은 TODO로 적어두면 좋을 것 같아요!
```suggestion
        val userId: Long = 1 // TODO : 임시 userId
```

jsoonworld replied:
좋네요! (할까 말까 고민했었는데,, 덕분에 잘 챙겨갑니다~ 😊)
TODO로 명시해두면 나중에 수정해야 할 부분을 더 분명하게 파악할 수 있어서 확실히 도움이 될 것 같아요.
말씀 주신 대로 // TODO: 임시 userId 형태로 반영하겠습니다! 감사합니다 🙏✨

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/auth/Auth.kt:
```diff
@@ -29,7 +29,7 @@ class Auth private constructor(
 
     @OneToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name = "userId", nullable = false)
+    @JoinColumn(name = "user_id", nullable = false)
```
카멜 케이스에서 스네이크 케이스로 바꾼 이유가 있나요?!

jsoonworld replied:
확인 감사합니다! 🙌
말씀 주신 부분처럼, 단순한 스타일 차이가 아니라 실제로 JPA 매핑 시 컬럼 이름 불일치로 인해 userId 대신 user_id로 명시해야만 동작하는 이슈가 있었습니다.

초기에는 userId로 사용했는데, 데이터베이스에서는 스네이크 케이스(user_id)로 컬럼이 생성되어 있어서 런타임에 매핑 오류가 발생했어요 😥
그래서 명시적으로 `@JoinColumn(name = "user_id")`로 수정해둔 상황이었습니다.

다만 이후 이슈를 해결하고 나서 다시 반영을 깜빡했었네요!
지금은 다시 userId로 되돌려 두었고,
말씀처럼 컨벤션 통일성과 명확한 기준을 가져가는 게 더 중요하다고 생각돼서, 앞으로도 그런 방향으로 관리해보겠습니다 :)
꼼꼼한 피드백 감사드립니다! 🙏

leeeyubin commented on build.gradle.kts:
```diff
@@ -37,13 +37,20 @@ dependencies {
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
     testImplementation("com.h2database:h2")
+
+    testImplementation("com.ninja-squad:springmockk:4.0.2")
+    testImplementation("io.mockk:mockk:1.13.10")
+    testImplementation("org.springframework.security:spring-security-test")
```
라이브러리 이름이 닌자(ninja)인게 신기하네요😂
모두 테스트를 위한 라이브러리 맞죠? 참고해서 사용하겠습니다!

jsoonworld replied:
맞아요! 모두 테스트 전용 라이브러리예요 😊
`springmockk`는 `Mockk`와 스프링 통합 테스트를 쉽게 해주는 라이브러리고,
`mockk`는 `Kotlin` 환경에서 사용하기 좋은 `mocking` 프레임워크예요.
`spring-security-test`는 시큐리티 설정이 적용된 환경에서 인증/인가 관련 테스트할 때 꼭 필요하더라고요!

저도 처음에 닌자 이름 보고 살짝 웃었어요 ㅎㅎ
테스트 작성하실 때 참고되면 좋겠습니다! 🥷✨

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/application/scrap/ScrapService.kt:
```diff
@@ -0,0 +1,45 @@
+
+@Service
+@Transactional(readOnly = false)
+class ScrapService(
```
`@Transactional`에 `readOnly = false`를 해 준 것은 기본 값이 true여서 그런가용?

jsoonworld replied:
이번에 트랜잭션 설정 관련 피드백을 받고 다시 점검해보면서,  
클래스 레벨에 `@Transactional(readOnly = false)`를 붙여둔 게  
제가 의도했던 방향과 다르다는 걸 발견했어요..! 레전드 실수 😅  
덕분에 바로잡을 수 있었습니다. 피드백 정말 감사합니다! 🙏

---

## 왜 클래스에 `@Transactional(readOnly = true)`를 붙이는가?

`@Transactional(readOnly = true)`를 선언하면 JPA 내부적으로 다음과 같은 최적화가 적용됩니다:

- **쓰기 지연(Write Delay)** 비활성화  
- **더티 체킹(Dirty Checking)** 비활성화  

즉, 불필요한 트랜잭션 관련 오버헤드를 줄일 수 있어  
**조회 성능을 향상시키는 데 유리**합니다.

---

##  어떻게 적용하고 있는가?

저는 보통 트랜잭션 설정을 다음과 같이 나누어 적용하고 있어요:

- **조회가 중심인 클래스**  
  → 클래스 레벨에 `@Transactional(readOnly = true)` 선언  

- **쓰기 작업(등록, 수정, 삭제 등)이 포함된 메서드**  
  → 해당 메서드에만 `@Transactional`을 붙여 기본값(readOnly = false)을 사용  

이렇게 하면 메서드 단위에서 의도가 명확해지고,  
전체적으로도 **구조가 깔끔**해지면서 **성능적인 이점**도 챙길 수 있습니다.

(보통 서비스 클래스에는 트랜잭션이 많이 발생하기 때문에 클레스 레벨에 readOnly를 걸어두고 가긴합니다.)

---

## ScrapService에 적용한 방식

`ScrapService` 클래스는 대부분 조회 로직으로 구성되어 있어  
클래스 레벨에 `@Transactional(readOnly = true)`를 선언했습니다.

다만 `scrap()` 메서드는 다음과 같은 **쓰기 작업**을 수행하기 때문에,  
해당 메서드에만 명시적으로 `@Transactional`을 붙였습니다:

- 스크랩 중복 여부 확인 및 예외 처리  
- 인턴십 공고 및 사용자 조회  
- **스크랩 객체 생성 및 저장**  
- **공고의 스크랩 수 증가**

---

## 정리하자면

- 클래스 레벨 → `@Transactional(readOnly = true)` : **조회 기본 전제**  
- 메서드 레벨 → `@Transactional` : **쓰기 의도 명확하게 전달**

이렇게 구분해두면 구조도 명확하고,  
무엇보다 의도를 코드에서 자연스럽게 읽을 수 있어  
**유지보수와 성능 측면 모두에서 효과적**이라고 생각합니다! 🙌

leeeyubin commented on src/test/kotlin/com/terning/server/kotlin/application/scrap/ScrapServiceTest.kt:
```diff
@@ -0,0 +1,100 @@
+    @Test
+    @DisplayName("이미 스크랩한 경우 예외가 발생한다")
+    fun scrapFailsIfAlreadyScrapped() {
+        every { scrapRepository.existsByInternshipAnnouncementIdAndUserId(userId, announcementId) } returns true
+
+        val exception =
+            assertThrows(ScrapException::class.java) {
+                scrapService.scrap(userId, announcementId, request)
+            }
```
every를  처음 봐서 조금 알아봤는데요!
every로 `existsByInternshipAnnouncementIdAndUserId`가 불리면 true가 반환된다는 것을 가정하고,
assertThrows로 실제 예외를 발생시켜서 확인을 하는 로직이 맞을까요..?

jsoonworld replied:
맞아요! 😊
말씀해주신 이해가 정확합니다!

여기서 every { ... } returns ... 구문은 mockk 라이브러리의 stubbing 방식인데요,
말 그대로 "이 메서드가 호출되면 이 값을 반환하라"는 행동(behavior)을 미리 정의하는 거예요.

예를 들어 아래 코드는:
```kotlin
every { scrapRepository.existsByInternshipAnnouncementIdAndUserId(userId, announcementId) } returns true
```
→ scrap() 호출 중에 existsByInternshipAnnouncementIdAndUserId()가 실행되면
그 결과를 true로 가정하고 동작하게끔 설정한 것이고요,
```kotlin
val exception = assertThrows<ScrapException> {
    scrapService.scrap(userId, announcementId, request)
}
```
→ 그 결과 실제로 ScrapException이 발생하는지를 검증하는 테스트입니다!

즉,

- given (이미 스크랩한 상태)

- when (스크랩을 시도하면)

- then (예외가 발생해야 한다)

이런 흐름을 가진 테스트라고 보시면 됩니다 🙌

mockk에 익숙하지 않더라도 위 구조만 기억하시면 금방 적응하실 수 있어요!
궁금한 점 또 생기면 편하게 말씀 주세요 😄

leeeyubin reviewed (approved):
API 구현하시느라 수고 많으셨어요!👍
보고 많이 배워갑니다!!

그리고 질문 주신 구조 관련해서도 고민을 해 보았어요! 
아직 여러 케이스들을 구현해 보지 않아서 어떤 게 더 나을지 확신이 안 서는 상황이긴 한데요.. 지금 생각나는 건 하나의 인터페이스 안에 여러 기능을 넣어서 관리하는 게 좋지 않을까 하는 생각은 있네요..!

jsoonworld committed:
refactor(ScrapService) : 네임드 아규먼트 사용으로 가독성 챙기기

- Scrap.of(...) 호출 시 네임드 아규먼트를 적용하여 인자의 의미를 명확히 표현했습니다.
- 협업 및 유지보수 시 가독성 향상을 위한 일관성 있는 코드 스타일을 반영했습니다.

jsoonworld committed:
refactor(UserName) : from 정적 팩토리 메서드를 표현식으로 변경

- 기존 블록 형태의 from 메서드를 표현식 본문으로 변경하여 간결하게 개선했습니다.

jsoonworld committed:
refactor(ScrapController) : 인증 미구현에 따른 임시 userId 주석 보완

jsoonworld committed:
refactor(Auth) : userId 컬럼명 스네이크 케이스 → 카멜 케이스로 수정

jsoonworld committed:
refactor(ScrapService) : 클래스 레벨 트랜잭션을 readOnly = true로 변경

jsoonworld commented:
> 그리고 질문 주신 구조 관련해서도 고민을 해 보았어요!
아직 여러 케이스들을 구현해 보지 않아서 어떤 게 더 나을지 확신이 안 서는 상황이긴 한데요.. 지금 생각나는 건 하나의 인터페이스 안에 여러 기능을 넣어서 관리하는 게 좋지 않을까 하는 생각은 있네요..!

이 부분은 API를 구현해가면서 코드 양이 많아지거나, 유지보수 과정에서 하나의 변경이 다른 코드에 영향을 많이 주게 되거나, 구조적으로 가시성이 떨어진다고 느껴질 때 다시 한 번 구조를 고민해봐도 좋을 것 같네요!
